File: Java/Structural/DecoratorPattern.java
package Structural;

// Component
interface Car {
    void assemble();
}

// Concrete Component
class BasicCar implements Car {
    @Override
    public void assemble() {
        System.out.println("Basic Car.");
    }
}

// Decorator
class CarDecorator implements Car {
    protected Car decoratedCar;
    
    public CarDecorator(Car car) {
        this.decoratedCar = car;
    }
    
    @Override
    public void assemble() {
        this.decoratedCar.assemble();
    }
}

// Concrete Decorators
class SportsCar extends CarDecorator {
    public SportsCar(Car car) {
        super(car);
    }
    
    @Override
    public void assemble() {
        super.assemble();
        System.out.println("Adding features of Sports Car.");
    }
}

class LuxuryCar extends CarDecorator {
    public LuxuryCar(Car car) {
        super(car);
    }
    
    @Override
    public void assemble() {
        super.assemble();
        System.out.println("Adding features of Luxury Car.");
    }
}

File: Python/Structural/DecoratorPattern.py
# Component
class Car:
    def assemble(self):
        pass

# Concrete Component
class BasicCar(Car):
    def assemble(self):
        print("Basic Car assembled")

# Decorator
class CarDecorator(Car):
    def __init__(self, car):
        self._car = car

    def assemble(self):
        self._car.assemble()

# Concrete Decorators
class SportsCar(CarDecorator):
    def assemble(self):
        self._car.assemble()
        print("Adding features of Sports Car")

class LuxuryCar(CarDecorator):
    def assemble(self):
        self._car.assemble()
        print("Adding features of Luxury Car")

# Client usage
basic_car = BasicCar()
sports_car = SportsCar(basic_car)
sports_car.assemble()

luxury_sports_car = LuxuryCar(sports_car)
luxury_sports_car.assemble()
